pymc3/step_methods/sgmcmc.py

```diff
@@ -0,0 +1,250 @@
+    M1 = 2147483647
+    M2 = 2147462579
+    return np.random.randint(0, M1, 3).tolist() + \
+        np.random.randint(0, M2, 3).tolist()
+
+
+def dlogp_elemwise_and_prior(vars, shared, model):
+    """Helper to create required tensors for SGFS. 
+    It returns the Jacobian of training data wrt model parameters
+    of size N x D. The gradient of log likelihood of the prior 
+    for the parameters and the inarray which is the input for the
+    two tensors
+    """
+
+    terms = [v.logp_elemwiset for v in model.observed_RVs]
+    elemwise_logp = tt.add(*terms)
+
+    def elemwise_dlogp_wrt_var(var):
+        """Returns elementwise gradient of log likelihood for each training datum wrt var"""
+        outputs, _ = theano.scan(\
+                    lambda i, elogp, v: theano.grad(elogp[i], v),\
+                    sequences=[theano.tensor.arange(elemwise_logp.shape[0])],
+                    non_sequences=[elemwise_logp, var])
+
+        return outputs
+
+    dlogp_elemwise = tt.as_tensor_variable(
+        [elemwise_dlogp_wrt_var(model, v) for v in vars]).T
+    dlog_prior = tt.as_tensor_variable([theano.grad(v.logpt, v) for v in vars])
+    sp_outputs, inarray = join_nonshared_inputs([dlogp_elemwise, dlog_prior],
+                                                vars, shared)
+    return sp_outputs[0], sp_outputs[1], inarray
+
+
+class BaseStochasticGradient(ArrayStepShared):
+    """
+    BaseStochasticGradient Object
+    
+    For working with BaseStochasticGradient Object
+    we need to supply the probabilistic model 
+    (:code:`model`) with the data supplied to observed
+    variables of type `GeneratorOp`
+    
+    Parameters
+    -------
+    vars : list
+        List of variables for sampler
+    batch_size`: int
+        Batch Size for each step
+    total_size : int
+        Total size of the training data
+    step_size : float
+        Step size for the parameter update
+    model : PyMC Model
+        Optional model for sampling step. Defaults to None (taken from context)
+    random_seed : int
+        The seed to initialize the Random Stream
+    """
+
+    def __init__(self,
+                 vars=None,
+                 batch_size=None,
+                 total_size=None,
+                 step_size=1.0,
+                 step_size_decay=100,
+                 model=None,
+                 random_seed=None,
+                 **kwargs):
+
+        model = modelcontext(model)
+        if vars is None:
+            vars = inputvars(model)
+
+        self.model = model
+        self.vars = vars
+        self.batch_size = batch_size
+        self.total_size = total_size
+        _value_error(
+            total_size != None or batch_size != None,
+            'total_size and batch_size of training data have to be specified')
+        self.expected_iter = int(total_size / batch_size)
+
+        # set random stream
+        self.random = None
+        if random_seed:
+            self.random = RandomStreams(random_seed)
```

Comment:
what about using `pm.theanof.tt_rng()`? in #2027 I add support for random seed if necessary. This will make codebase a bit more consistent.